pkg/controller/replication_controller.go

```diff
@@ -115,181 +91,276 @@ func NewReplicationManager(kubeClient client.Interface) *ReplicationManager {
+		},
+	)
+
+	rm.podStore.Store, rm.podController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return rm.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return rm.kubeClient.Pods(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Pod{},
+		PodRelistPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: rm.addPod,
+			// This invokes the rc for every pod change, eg: host assignment. Though this might seem like overkill
+			// the most frequent pod update is status, and the associated rc will only list from local storage, so
+			// it should be ok.
+			UpdateFunc: rm.updatePod,
+			DeleteFunc: rm.deletePod,
+		},
+	)
+
 	rm.syncHandler = rm.syncReplicationController
 	return rm
 }
 
 // Run begins watching and syncing.
-func (rm *ReplicationManager) Run(period time.Duration) {
-	rm.syncTime = time.Tick(period)
-	resourceVersion := ""
-	go util.Forever(func() { rm.watchControllers(&resourceVersion) }, period)
+func (rm *ReplicationManager) Run(workers int, stopCh <-chan struct{}) {
+	defer util.HandleCrash()
+	go rm.rcController.Run(stopCh)
+	go rm.podController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go util.Until(rm.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	rm.queue.ShutDown()
 }
 
-// resourceVersion is a pointer to the resource version to use/update.
-func (rm *ReplicationManager) watchControllers(resourceVersion *string) {
-	watching, err := rm.kubeClient.ReplicationControllers(api.NamespaceAll).Watch(
-		labels.Everything(),
-		fields.Everything(),
-		*resourceVersion,
-	)
+// getPodControllers returns a set of keys describing the controllers that control a given pod.
+func (rm *ReplicationManager) getPodControllers(pod *api.Pod) (map[string]api.ReplicationController, error) {
+	controllerLookup := map[string]api.ReplicationController{}
+	if pod.Labels == nil {
+		// Pods without labels match nothing, not everything.
+		return controllerLookup, nil
+	}
+	controllers, err := rm.controllerStore.Controllers(pod.Namespace).List(labels.Set(pod.Labels).AsSelector())
+	if err != nil {
+		// TODO: Surface this error
+		return controllerLookup, nil
+	}
+	// TODO: Multiple rcs for a given pod is unsupported, currently they will fight it out.
+	for _, rc := range controllers {
+		key, err := rcKeyFunc(&rc)
+		if err != nil {
+			return nil, err
+		}
+		controllerLookup[key] = rc
+	}
+	return controllerLookup, nil
+}
+
+func (rm *ReplicationManager) addPod(obj interface{}) {
+	pod := obj.(*api.Pod)
+	controllers, err := rm.getPodControllers(pod)
 	if err != nil {
-		util.HandleError(fmt.Errorf("unable to watch: %v", err))
-		time.Sleep(5 * time.Second)
+		glog.Errorf("Unable to get pod %v/%v's controllers: %v", pod.Namespace, pod.Name, err)
 		return
 	}
+	for key, rc := range controllers {
+		rm.expectations.LowerExpectations(rc, 1, 0)
```

Comment:
You can make a private method to reduce code duplication. Interface clarity is really important.